Source/WebCore/style/StyleTreeResolver.h

```diff
@@ -160,22 +163,31 @@ class TreeResolver {
 
     bool hasUnresolvedAnchorPosition(const Element&) const;
 
-    struct QueryContainerState {
-        OptionSet<Change> changes;
-        DescendantsToResolve descendantsToResolve { DescendantsToResolve::None };
-        bool invalidated { false };
-    };
-
     CheckedRef<Document> m_document;
     std::unique_ptr<RenderStyle> m_documentElementStyle;
 
     Vector<Ref<Scope>, 4> m_scopeStack;
     Vector<Parent, 32> m_parentStack;
     bool m_didSeePendingStylesheet { false };
 
-    UncheckedKeyHashMap<Ref<Element>, std::optional<QueryContainerState>> m_queryContainerStates;
+    // For elements that require style/layout interleaving (anchor positioning and
+    // container queries), the first style pass will defer resolving descendants.
+    // This is resumed in subsequent style passes, when there are enough information
+    // to do so. This struct preserves internal resolution state when descendant
+    // resolution is deferred, so it could be restores when descendant resolution
+    // resumes.
+    struct StyleLayoutInterleavedElementState {
+        OptionSet<Change> changes;
+        DescendantsToResolve descendantsToResolve { DescendantsToResolve::None };
+    };
+    UncheckedKeyHashMap<Ref<Element>, StyleLayoutInterleavedElementState> m_styleLayoutInterleavedElementStates;
     bool m_needsInterleavedLayout { false };
 
+    struct QueryContainerState {
+        bool invalidated { false };
+    };
+    UncheckedKeyHashMap<Ref<Element>, QueryContainerState> m_queryContainerStates;
```

Comment:
Maybe there is a way to merge this state too with the interleaved state in future.